lib/navigation/preprocess_encoders.cpp

```diff
@@ -34,7 +38,38 @@ core::EncoderData EncodersPreprocessor::detectOutliers(const core::EncoderData e
   -also figure out return type/ what we update and update
   documentation as appropriate
   */
-  return {0, 0, 0, 0};
+ core::EncoderData encoder_data_copy = std::copy(encoder_data.begin(), encoder_data.end(), std::back_inserter(encoder_data));
+ std::sort(encoder_data_copy.begin(), encoder_data_copy.end());
+ core::Float median = 0;
+ core::Float q1 = 0.0;
+ core::Float q3 = 0.0;
+ if(encoder_data_copy.size() % 2 == 0){
+  median = (encoder_data_copy[(std::size(encoder_data_copy))/2] + encoder_data_copy[((std::size(encoder_data_copy))/2) + 1])/2.0;
+ }
+ else{
+  median = encoder_data_copy[(std::size(encoder_data_copy) + 1)/2];
+ }
+ core::EncoderData array_for_q1 ;
+ std::copy_if(encoder_data_copy.begin(),encoder_data_copy.end(),array_for_q1.begin(),[](uint32_t i){return i< median;});
+ core::EncoderData array_for_q2;
+ std::copy_if(encoder_data_copy.begin(),encoder_data_copy.end(),array_for_q2.begin(),[](uint32_t i){return i> median;});
+ if(array_for_q1.size() % 2 == 0){
+  q1 = (array_for_q1[(array_for_q1.size())/2] +array_for_q1[((array_for_q1.size())/2) + 1] )/2.0;
```

Comment:
lots of if statements and error handling here just to get upper and lower bounds dependent on if we have an unreliable encoder. Consider defining a function that takes in a copy of the encoder data and returns the median and the upper and lower bounds?

Reply:
This will help to avoid writing single functions that are too long. A function in general should not exceed 60 lines of code.